enhancements/vsphere-multi-vcenter.md

```diff
@@ -0,0 +1,839 @@
+need to update the permissions / roles of the vSphere CSI Driver Operator.
+
+#### vSphere CSI Driver Operator
+
+The vSphere CSI Driver operator has a lot of enhancements done for multi vCenter
+support.  These include:
+- Added feature gate for multi vCenter
+- Moved password from env variables for image to the CSI config file
+- Enhanced CSI config file to define each vCenter with user/pass
+- Updated check to include a connection for each vCenter
+- Created new wrapper config object to contain legacy config values when detected
+
+The important enhancements to discuss revolve around the change to config for csi 
+driver, using upstream config and improvement to all the checks.  The first topic
+of interest is the change to CSI config.  
+
+Today we are putting only one vCenter into the csi driver INI file.
+We are enhancing the process of generating that to now use a template that will 
+insert multiple vCenters.  While we are now adding each vCenter into this config 
+file, we are also going to be moving the username and password into the INI file 
+as well.  The env variables only really works for single vCenter, but does not 
+work for multiple.  Upstream has the ability to set all these values in the INI 
+config and so we are migrating to putting these there.
+
+For this to follow the security pattern of user/pass being in secrets and not
+config maps, we need to make sure that the csi driver INI config is moved from
+the configmap location it is currently using into a secret.  There was a separate
+PR that already exists that is in the process of doing this, so we will piggy-back
+off that PR to get this behavior in place.
+
+An example of the INI file w/ user and password configured:
+
+```ini
+# Labels with topology values are added dynamically via operator
+[Global]
+cluster-id = ngirard-multi-bcw8t
+
+# Populate VCenters (multi) after here
+[VirtualCenter "vcs8e-vc.ocp2.dev.cluster.com"]
+insecure-flag           = true
+datacenters             = IBMCloud
+migration-datastore-url = ds:///vmfs/volumes/vsan:523ea352e875627d-b090c96b526bb79c/
+password                = password
+user                    = user
+
+[VirtualCenter "vcenter.ci.ibmc.devcluster.openshift.com"]
+insecure-flag           = true
+datacenters             = cidatacenter
+migration-datastore-url = ds:///vmfs/volumes/vsan:523ea352e875627d-b090c96b526bb79c/
+password                = password2
+user                    = user2
+
+[Labels]
+topology-categories = openshift-zone,openshift-region
+```
+
+Next the operator was enhanced to be able to support using the upstream vSphere 
+YAML cloud provider config format.  There is some logic that uses our old legacy 
+style config.  To preserve this, we created a wrapper config object that attempts
+to load the cloud provider config as either INI or YAML.  If its INI, we will 
+also store the INI data into the `LegacyConfig` field so we can access it in 
+certain situations.
```

Comment:
We have had people confused by the errors from trying to read old vs new, is there anything we can do to mitigate these errors from being produced?

Reply:
I am working with upstream to fix these outputs: https://github.com/kubernetes/cloud-provider-vsphere/pull/1225